DESIGN.md

```diff
@@ -0,0 +1,133 @@
+
+We will make changes to the parser, binder, optimizer, planner, and execution layers, to support different types of ALTER TABLE SQL commands.
+
+## Architectural Design
+
+### SqlTable
+
+To manage multiple datatables, each Sqltable keeps a ordered map (tables_) from version number to a struct containing the metadata of each datatable (called DataTableVersion). 
+
+Note that in the vast majority of cases, there will be no schema changes, and each Sqltable manages only one Datatable. If we need to go through tables_, which is a concurrent map (or vector), on each call to Sqltable, that will be a huge overhead. Therefore, we will cache the DataTable in the SqlTable when there is only one DataTable to bypass the map/vector lookup.
+
+Each Sqltable has a separate version number counter, managed by the catalog. Each Sqltable starts out with version number 0, and manages a single datatable. When UpdateSchema() is called, the version number is incremented, and we create a new datatable via SqlTable::CreateTable(), which is the intended location for all tuples under the new schema. Note that within a Sqltable, each datatable uniquely corresponds to a version number and a schema, both of which never change. Therefore, in each datatable, we store its version number.  Moreover, for each datatable within Sqltable,  we store in DataTableVersion its schema, a default value map, and two maps from column oid to id and from id to oid.
+
+To do an SQL query such as updating a tuple, a transaction first uses the index on Sqltable to find the tupleslot of the tuple (note that we update the index when migrating tupeslots). We can get the raw block from the top 44 bits of the tupleslot, and the raw block has reference to the datatable of the tupleslot.   Thus we know the version number of the datatable the tuple is current stored in, call this the *storage_version*.
+
+Each SQL query also comes with an argument, version_number, which is the schema version of the Sqltable that should be seen by the current transaction. Let us call these versions *intended_version*. 
+
+To migrate a tuple with primary key *K* from an older datatable to a newer datatable, we first logically delete the tuple from the older datatable with the current transaction's timestamp, then insert the tuple to the newer datatable. Note that the old tupleslot for *K* has been deleted and a new tupleslot was created. We will update the index by logically deleting *K* then inserting *K* with the new tupleslot.
+
+Note that *storage_version* might be smaller than *intended_version*, which occurs if a transaction accesses a tuple of a newer schema version, but the tuple has not yet been migrated over from old datatables. Different Sqltable operations deal with this special case of  *storage_version < intended_version* differently, as we will  describe below.
```

Comment:
For unique transactions with timestamps t1, t2, t2 > t1, does the transaction with timestamp t2 have an intended_version greater than the transaction with timestamp t1? If so, wouldn't the case of having storage_version > intended_version need to be considered?

Reply:
For example, if the transaction with timestamp t2 commits before the transaction with timestamp t1, storage_version would be updated to a version number higher than the intended_version associated with the transaction with timestamp t1.

Reply:
I think you are mixing up versioning with transaction timestamps. They are 2 completely different things.

Version only updates when updateschema is called. For any tupleslot, it will either be placed in the datatable with correct version, or be in a datatable with older version. We will never more a tupleslot to a datatable with version newer then its intended version.